source/common/network/utility.cc

```diff
@@ -542,11 +539,71 @@ void passPayloadToProcessor(uint64_t bytes_read, Buffer::RawSlice& slice,
                                      std::move(buffer), receive_time);
 }
 
+Api::IoCallUint64Result receiveMessage(uint64_t max_packet_size, Buffer::InstancePtr& buffer,
+                                       Buffer::RawSlice& slice, IoHandle::RecvMsgOutput& output,
+                                       IoHandle& handle, const Address::Instance& local_address) {
+
+  const uint64_t num_slices = buffer->reserve(max_packet_size, &slice, 1);
+  ASSERT(num_slices == 1u);
+
+  Api::IoCallUint64Result result =
+      handle.recvmsg(&slice, num_slices, local_address.ip()->port(), output);
+
+  if (!result.ok()) {
+    return result;
+  }
+
+  // Adjust memory length and commit slice to buffer
+  slice.len_ = std::min(slice.len_, static_cast<size_t>(result.rc_));
+  buffer->commit(&slice, 1);
+
+  return result;
+}
+
 Api::IoCallUint64Result Utility::readFromSocket(IoHandle& handle,
                                                 const Address::Instance& local_address,
                                                 UdpPacketProcessor& udp_packet_processor,
                                                 MonotonicTime receive_time,
                                                 uint32_t* packets_dropped) {
+
+  if (handle.supportsUdpGro()) {
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    Buffer::RawSlice slice;
+    IoHandle::RecvMsgOutput output(1, packets_dropped);
+
+    // TODO(yugant): Avoid using 64k by getting memory from UdpPacketProcessor
+    const uint64_t max_packet_size_with_gro = 65535;
```

Comment:
nit: change to `16 * udp_packet_processor.maxPacketSize()`

Reply:
Done

Reply:
Using 16 here.
Updated the comment above to say 24k instead of 64k.